### ISSUE TITLE
add_distances_ca issue

### ISSUE THREAD

**Issue by euhruska** [markovmodel/PyEMMA#735]
Hi, the add_distances_ca() is giving me an error I can't explain. What happened? Is my gro file wrong?

```
>>>feat = pyemma.coordinates.featurizer('start.gro')
>>>feat.add_distances_ca()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pyemma/coordinates/data/featurization/featurizer.py", line 285, in add_distances_ca
    distance_indexes = ca_at_idxs[distance_indexes]
IndexError: index 12 is out of bounds for axis 1 with size 12
```

If I use another gro file it works but with the following one not:

```
Generated by trjconv : Protein t=   0.00000
  132
    1ACE   HH31    1  50.3645  50.6820  50.2925
    1ACE    CH3    2  50.3665  50.7360  50.3875
    1ACE   HH32    3  50.4705  50.7640  50.4045
    1ACE   HH33    4  50.3095  50.8290  50.3925
    1ACE      C    5  50.3095  50.6410  50.4915
    1ACE      O    6  50.1885  50.6230  50.4845
    2ALA      N    7  50.3955  50.5620  50.5555
    2ALA      H    8  50.4925  50.5840  50.5355
    2ALA     CA    9  50.3595  50.4470  50.6365
    2ALA     HA   10  50.4415  50.3760  50.6345
    2ALA     CB   11  50.3305  50.4880  50.7805
    2ALA    HB1   12  50.3155  50.3930  50.8325
    2ALA    HB2   13  50.2615  50.5710  50.7925
    2ALA    HB3   14  50.4255  50.5190  50.8255
    2ALA      C   15  50.2465  50.3660  50.5735
    2ALA      O   16  50.2815  50.2990  50.4755
    3ALA      N   17  50.1275  50.3680  50.6335
    3ALA      H   18  50.1235  50.4360  50.7075
    3ALA     CA   19  50.0155  50.2810  50.6015
    3ALA     HA   20  50.0495  50.1790  50.6165
    3ALA     CB   21  49.8985  50.3050  50.6955
    3ALA    HB1   22  49.8615  50.4070  50.6785
    3ALA    HB2   23  49.9365  50.2950  50.7965
    3ALA    HB3   24  49.8155  50.2370  50.6755
    3ALA      C   25  49.9805  50.3040  50.4555
    3ALA      O   26  49.9825  50.2040  50.3835
    4ALA      N   27  49.9695  50.4290  50.4115
    4ALA      H   28  50.0065  50.5000  50.4735
    4ALA     CA   29  49.9335  50.4620  50.2745
    4ALA     HA   30  49.8265  50.4430  50.2635
    4ALA     CB   31  49.9305  50.6140  50.2645
    4ALA    HB1   32  49.8715  50.6540  50.3465
    4ALA    HB2   33  49.8825  50.6390  50.1695
    4ALA    HB3   34  50.0305  50.6590  50.2575
    4ALA      C   35  50.0255  50.3980  50.1715
    4ALA      O   36  49.9765  50.3430  50.0735
    5ALA      N   37  50.1555  50.4130  50.1975
    5ALA      H   38  50.1775  50.4620  50.2835
    5ALA     CA   39  50.2605  50.3600  50.1115
    5ALA     HA   40  50.2445  50.3960  50.0105
    5ALA     CB   41  50.3965  50.4160  50.1535
    5ALA    HB1   42  50.4715  50.3850  50.0805
    5ALA    HB2   43  50.4265  50.3770  50.2505
    5ALA    HB3   44  50.3935  50.5250  50.1525
    5ALA      C   45  50.2625  50.2080  50.1095
    5ALA      O   46  50.2605  50.1510  50.0005
    6ALA      N   47  50.2655  50.1390  50.2235
    6ALA      H   48  50.2775  50.1890  50.3105
    6ALA     CA   49  50.2495  49.9950  50.2345
    6ALA     HA   50  50.3315  49.9420  50.1845
    6ALA     CB   51  50.2535  49.9410  50.3765
    6ALA    HB1   52  50.2365  49.8330  50.3705
    6ALA    HB2   53  50.1845  49.9860  50.4485
    6ALA    HB3   54  50.3515  49.9570  50.4225
    6ALA      C   55  50.1205  49.9500  50.1675
    6ALA      O   56  50.1205  49.8590  50.0855
    7ALA      N   57  50.0035  50.0070  50.1975
    7ALA      H   58  49.9995  50.0700  50.2755
    7ALA     CA   59  49.8725  49.9750  50.1445
    7ALA     HA   60  49.8425  49.8730  50.1725
    7ALA     CB   61  49.7775  50.0710  50.2155
    7ALA    HB1   62  49.8115  50.1730  50.1985
    7ALA    HB2   63  49.7645  50.0520  50.3215
    7ALA    HB3   64  49.6795  50.0510  50.1715
    7ALA      C   65  49.8695  49.9870  49.9935
    7ALA      O   66  49.8505  49.8860  49.9265
    8ALA      N   67  49.8925  50.1080  49.9415
    8ALA      H   68  49.9215  50.1840  50.0005
    8ALA     CA   69  49.9125  50.1210  49.7985
    8ALA     HA   70  49.8085  50.1140  49.7645
    8ALA     CB   71  49.9565  50.2640  49.7665
    8ALA    HB1   72  50.0515  50.2880  49.8135
    8ALA    HB2   73  49.8855  50.3420  49.7925
    8ALA    HB3   74  49.9815  50.2640  49.6605
    8ALA      C   75  50.0095  50.0260  49.7305
    8ALA      O   76  49.9845  49.9870  49.6165
    9ALA      N   77  50.1245  49.9960  49.7915
    9ALA      H   78  50.1565  50.0560  49.8665
    9ALA     CA   79  50.2145  49.8950  49.7405
    9ALA     HA   80  50.2355  49.9150  49.6355
    9ALA     CB   81  50.3445  49.8930  49.8215
    9ALA    HB1   82  50.3275  49.8660  49.9255
    9ALA    HB2   83  50.3905  49.9910  49.8335
    9ALA    HB3   84  50.4175  49.8240  49.7795
    9ALA      C   85  50.1515  49.7560  49.7385
    9ALA      O   86  50.1725  49.6810  49.6435
   10ALA      N   87  50.0745  49.7240  49.8425
   10ALA      H   88  50.0485  49.7880  49.9155
   10ALA     CA   89  50.0235  49.5890  49.8605
   10ALA     HA   90  50.0995  49.5160  49.8335
   10ALA     CB   91  49.9905  49.5640  50.0075
   10ALA    HB1   92  50.0775  49.5780  50.0705
   10ALA    HB2   93  49.9575  49.4620  50.0275
   10ALA    HB3   94  49.9185  49.6360  50.0465
   10ALA      C   95  49.9095  49.5550  49.7665
   10ALA      O   96  49.9105  49.4510  49.7015
   11ALA      N   97  49.8105  49.6450  49.7595
   11ALA      H   98  49.8065  49.7180  49.8295
   11ALA     CA   99  49.7095  49.6550  49.6555
   11ALA     HA  100  49.6475  49.5650  49.6505
   11ALA     CB  101  49.6035  49.7460  49.7175
   11ALA    HB1  102  49.6145  49.8530  49.7025
   11ALA    HB2  103  49.5945  49.7210  49.8235
   11ALA    HB3  104  49.5075  49.7120  49.6805
   11ALA      C  105  49.7585  49.7010  49.5195
   11ALA      O  106  49.7005  49.7930  49.4615
   12ALA      N  107  49.8685  49.6390  49.4745
   12ALA      H  108  49.8995  49.5710  49.5415
   12ALA     CA  109  49.9325  49.6570  49.3465
   12ALA     HA  110  49.8485  49.6420  49.2785
   12ALA     CB  111  49.9825  49.8000  49.3315
   12ALA    HB1  112  50.0435  49.8080  49.2415
   12ALA    HB2  113  50.0385  49.8250  49.4215
   12ALA    HB3  114  49.9065  49.8760  49.3135
   12ALA      C  115  50.0445  49.5590  49.3135
   12ALA      O  116  50.0295  49.4690  49.2305
   13ALA      N  117  50.1555  49.5660  49.3855
   13ALA      H  118  50.1585  49.6210  49.4705
   13ALA     CA  119  50.2815  49.5020  49.3495
   13ALA     HA  120  50.2625  49.4120  49.2905
   13ALA     CB  121  50.3555  49.6010  49.2605
   13ALA    HB1  122  50.4525  49.5570  49.2345
   13ALA    HB2  123  50.3795  49.6910  49.3165
   13ALA    HB3  124  50.3015  49.6170  49.1675
   13ALA      C  125  50.3525  49.4660  49.4795
   13ALA      O  126  50.4325  49.5470  49.5265
   14NME      N  127  50.3285  49.3450  49.5295
   14NME      H  128  50.2495  49.2930  49.4955
   14NME    CH3  129  50.3935  49.2790  49.6415
   14NME   HH31  130  50.3255  49.3020  49.7235
   14NME   HH32  131  50.4905  49.3260  49.6565
   14NME   HH33  132  50.3955  49.1710  49.6315
 100.00000 100.00000 100.00000
```

**Comment by gph82** [comment]
Hi, I will look into this soon, thanks!

**Comment by franknoe** [comment]
Are we assuming that every protein residue has a CA? That would explain 
that error, because in Eugen's gro file the termini are in separate 
residues.

Am 17/03/16 um 18:19 schrieb Eugen Hruska:

> Hi, the add_distances_ca() is giving me an error I can't explain. What 
> happened? Is my gro file wrong?
> 
> |>>>feat = pyemma.coordinates.featurizer('start.gro') 
> 
> > > > feat.add_distances_ca() Traceback (most recent call last): File 
> > > > "<stdin>", line 1, in <module> File 
> > > > "/usr/local/lib/python2.7/site-packages/pyemma/coordinates/data/featurization/featurizer.py", 
> > > > line 285, in add_distances_ca distance_indexes = 
> > > > ca_at_idxs[distance_indexes] IndexError: index 12 is out of bounds for 
> > > > axis 1 with size 12 |
> 
> If I use another gro file it works but with the following one not:
> 
> |Generated by trjconv : Protein t= 0.00000 132 1ACE HH31 1 50.3645 
> 50.6820 50.2925 1ACE CH3 2 50.3665 50.7360 50.3875 1ACE HH32 3 50.4705 
> 50.7640 50.4045 1ACE HH33 4 50.3095 50.8290 50.3925 1ACE C 5 50.3095 
> 50.6410 50.4915 1ACE O 6 50.1885 50.6230 50.4845 2ALA N 7 50.3955 
> 50.5620 50.5555 2ALA H 8 50.4925 50.5840 50.5355 2ALA CA 9 50.3595 
> 50.4470 50.6365 2ALA HA 10 50.4415 50.3760 50.6345 2ALA CB 11 50.3305 
> 50.4880 50.7805 2ALA HB1 12 50.3155 50.3930 50.8325 2ALA HB2 13 
> 50.2615 50.5710 50.7925 2ALA HB3 14 50.4255 50.5190 50.8255 2ALA C 15 
> 50.2465 50.3660 50.5735 2ALA O 16 50.2815 50.2990 50.4755 3ALA N 17 
> 50.1275 50.3680 50.6335 3ALA H 18 50.1235 50.4360 50.7075 3ALA CA 19 
> 50.0155 50.2810 50.6015 3ALA HA 20 50.0495 50.1790 50.6165 3ALA CB 21 
> 49.8985 50.3050 50.6955 3ALA HB1 22 49.8615 50.4070 50.6785 3ALA HB2 
> 23 49.9365 50.2950 50.7965 3ALA HB3 24 49.8155 50.2370 50.6755 3ALA C 
> 25 49.9805 50.3040 50.4555 3ALA O 26 49.9825 50.2040 50.3835 4ALA N 27 
> 49.9695 50.4290 50.4115 4ALA H 28 50.0065 50.5000 50.4735 4ALA CA 29 
> 49.9335 50.4620 50.2745 4ALA HA 30 49.8265 50.4430 50.2635 4ALA CB 31 
> 49.9305 50.6140 50.2645 4ALA HB1 32 49.8715 50.6540 50.3465 4ALA HB2 
> 33 49.8825 50.6390 50.1695 4ALA HB3 34 50.0305 50.6590 50.2575 4ALA C 
> 35 50.0255 50.3980 50.1715 4ALA O 36 49.9765 50.3430 50.0735 5ALA N 37 
> 50.1555 50.4130 50.1975 5ALA H 38 50.1775 50.4620 50.2835 5ALA CA 39 
> 50.2605 50.3600 50.1115 5ALA HA 40 50.2445 50.3960 50.0105 5ALA CB 41 
> 50.3965 50.4160 50.1535 5ALA HB1 42 50.4715 50.3850 50.0805 5ALA HB2 
> 43 50.4265 50.3770 50.2505 5ALA HB3 44 50.3935 50.5250 50.1525 5ALA C 
> 45 50.2625 50.2080 50.1095 5ALA O 46 50.2605 50.1510 50.0005 6ALA N 47 
> 50.2655 50.1390 50.2235 6ALA H 48 50.2775 50.1890 50.3105 6ALA CA 49 
> 50.2495 49.9950 50.2345 6ALA HA 50 50.3315 49.9420 50.1845 6ALA CB 51 
> 50.2535 49.9410 50.3765 6ALA HB1 52 50.2365 49.8330 50.3705 6ALA HB2 
> 53 50.1845 49.9860 50.4485 6ALA HB3 54 50.3515 49.9570 50.4225 6ALA C 
> 55 50.1205 49.9500 50.1675 6ALA O 56 50.1205 49.8590 50.0855 7ALA N 57 
> 50.0035 50.0070 50.1975 7ALA H 58 49.9995 50.0700 50.2755 7ALA CA 59 
> 49.8725 49.9750 50.1445 7ALA HA 60 49.8425 49.8730 50.1725 7ALA CB 61 
> 49.7775 50.0710 50.2155 7ALA HB1 62 49.8115 50.1730 50.1985 7ALA HB2 
> 63 49.7645 50.0520 50.3215 7ALA HB3 64 49.6795 50.0510 50.1715 7ALA C 
> 65 49.8695 49.9870 49.9935 7ALA O 66 49.8505 49.8860 49.9265 8ALA N 67 
> 49.8925 50.1080 49.9415 8ALA H 68 49.9215 50.1840 50.0005 8ALA CA 69 
> 49.9125 50.1210 49.7985 8ALA HA 70 49.8085 50.1140 49.7645 8ALA CB 71 
> 49.9565 50.2640 49.7665 8ALA HB1 72 50.0515 50.2880 49.8135 8ALA HB2 
> 73 49.8855 50.3420 49.7925 8ALA HB3 74 49.9815 50.2640 49.6605 8ALA C 
> 75 50.0095 50.0260 49.7305 8ALA O 76 49.9845 49.9870 49.6165 9ALA N 77 
> 50.1245 49.9960 49.7915 9ALA H 78 50.1565 50.0560 49.8665 9ALA CA 79 
> 50.2145 49.8950 49.7405 9ALA HA 80 50.2355 49.9150 49.6355 9ALA CB 81 
> 50.3445 49.8930 49.8215 9ALA HB1 82 50.3275 49.8660 49.9255 9ALA HB2 
> 83 50.3905 49.9910 49.8335 9ALA HB3 84 50.4175 49.8240 49.7795 9ALA C 
> 85 50.1515 49.7560 49.7385 9ALA O 86 50.1725 49.6810 49.6435 10ALA N 
> 87 50.0745 49.7240 49.8425 10ALA H 88 50.0485 49.7880 49.9155 10ALA CA 
> 89 50.0235 49.5890 49.8605 10ALA HA 90 50.0995 49.5160 49.8335 10ALA 
> CB 91 49.9905 49.5640 50.0075 10ALA HB1 92 50.0775 49.5780 50.0705 
> 10ALA HB2 93 49.9575 49.4620 50.0275 10ALA HB3 94 49.9185 49.6360 
> 50.0465 10ALA C 95 49.9095 49.5550 49.7665 10ALA O 96 49.9105 49.4510 
> 49.7015 11ALA N 97 49.8105 49.6450 49.7595 11ALA H 98 49.8065 49.7180 
> 49.8295 11ALA CA 99 49.7095 49.6550 49.6555 11ALA HA 100 49.6475 
> 49.5650 49.6505 11ALA CB 101 49.6035 49.7460 49.7175 11ALA HB1 102 
> 49.6145 49.8530 49.7025 11ALA HB2 103 49.5945 49.7210 49.8235 11ALA 
> HB3 104 49.5075 49.7120 49.6805 11ALA C 105 49.7585 49.7010 49.5195 
> 11ALA O 106 49.7005 49.7930 49.4615 12ALA N 107 49.8685 49.6390 
> 49.4745 12ALA H 108 49.8995 49.5710 49.5415 12ALA CA 109 49.9325 
> 49.6570 49.3465 12ALA HA 110 49.8485 49.6420 49.2785 12ALA CB 111 
> 49.9825 49.8000 49.3315 12ALA HB1 112 50.0435 49.8080 49.2415 12ALA 
> HB2 113 50.0385 49.8250 49.4215 12ALA HB3 114 49.9065 49.8760 49.3135 
> 12ALA C 115 50.0445 49.5590 49.3135 12ALA O 116 50.0295 49.4690 
> 49.2305 13ALA N 117 50.1555 49.5660 49.3855 13ALA H 118 50.1585 
> 49.6210 49.4705 13ALA CA 119 50.2815 49.5020 49.3495 13ALA HA 120 
> 50.2625 49.4120 49.2905 13ALA CB 121 50.3555 49.6010 49.2605 13ALA HB1 
> 122 50.4525 49.5570 49.2345 13ALA HB2 123 50.3795 49.6910 49.3165 
> 13ALA HB3 124 50.3015 49.6170 49.1675 13ALA C 125 50.3525 49.4660 
> 49.4795 13ALA O 126 50.4325 49.5470 49.5265 14NME N 127 50.3285 
> 49.3450 49.5295 14NME H 128 50.2495 49.2930 49.4955 14NME CH3 129 
> 50.3935 49.2790 49.6415 14NME HH31 130 50.3255 49.3020 49.7235 14NME 
> HH32 131 50.4905 49.3260 49.6565 14NME HH33 132 50.3955 49.1710 
> 49.6315 100.00000 100.00000 100.00000 |
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/735

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by gph82** [comment]
Yes I am looking into that right now

**Comment by gph82** [comment]
@euhruska  thanks for noticing. It's now fixed with PR #737

``` python
import pyemma.coordinates as coor
feat = coor.featurizer('/home/mi/gph82/start.gro')
feat.add_distances_ca()
# Check
for ii, jj in feat.active_features[0].distance_indexes:
    print(feat.topology.atom(ii), feat.topology.atom(jj))
```

output:

```
ALA2-CA ALA5-CA
ALA2-CA ALA6-CA
ALA2-CA ALA7-CA
ALA2-CA ALA8-CA
ALA2-CA ALA9-CA
ALA2-CA ALA10-CA
ALA2-CA ALA11-CA
ALA2-CA ALA12-CA
ALA2-CA ALA13-CA
ALA3-CA ALA6-CA
ALA3-CA ALA7-CA
ALA3-CA ALA8-CA
ALA3-CA ALA9-CA
ALA3-CA ALA10-CA
ALA3-CA ALA11-CA
ALA3-CA ALA12-CA
ALA3-CA ALA13-CA
ALA4-CA ALA7-CA
ALA4-CA ALA8-CA
ALA4-CA ALA9-CA
ALA4-CA ALA10-CA
ALA4-CA ALA11-CA
ALA4-CA ALA12-CA
ALA4-CA ALA13-CA
ALA5-CA ALA8-CA
ALA5-CA ALA9-CA
ALA5-CA ALA10-CA
ALA5-CA ALA11-CA
ALA5-CA ALA12-CA
ALA5-CA ALA13-CA
ALA6-CA ALA9-CA
ALA6-CA ALA10-CA
ALA6-CA ALA11-CA
ALA6-CA ALA12-CA
ALA6-CA ALA13-CA
ALA7-CA ALA10-CA
ALA7-CA ALA11-CA
ALA7-CA ALA12-CA
ALA7-CA ALA13-CA
ALA8-CA ALA11-CA
ALA8-CA ALA12-CA
ALA8-CA ALA13-CA
ALA9-CA ALA12-CA
ALA9-CA ALA13-CA
ALA10-CA ALA13-CA
```
